### PR TITLE
[WIP] Master & Slave folders

### PR DESCRIPTION
**Pull request by Tracerneo**
Restructures and cleans up folder structs, renames functions and fields.
From compatibility breaking stuff, it replaces *ro* config field with *master* and *slave* fields.

Particular master/slave setting combinations should work as described:
* both false – normal push&pull folder, keep in sync
* master true – push-only folder, not accepting changesets
* slave true – pull-only folder, copy pushing nodes, always override local changes
* both true – no push nor pull, locally inform about sync status, possibly allow downloading as a mirror

Force Push and Force Pull could be added and act alike Override Changes on current ReadOnly (Master) folders.

Todo:
- [x] rename config field *ReadOnly* to *Master*
- [x] add *Slave* config field
- [x] prepare code for new folder types
- [ ] add Slave folders
- [ ] add Snapshot (Master+Slave) folders
- [ ] add Force Pull and Force Push
- [ ] update web interface
- [ ] update docs

Should solve #62 and #1439
I'm still working on it, not yet to be merged.

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
So so far so good, my only concern is changing usage reporting data, which we need to make sure its backwards compatible.

**Comment by calmh** [comment]
:+1:

Changing the config format is fine, as long as we introduce a new config version and a migration function for it.

**Comment by calmh** [comment]
@st-jenkins add to whitelist

**Comment by 1138-4EB** [comment]
Coming from [this thread](https://forum.syncthing.net/t/aditional-repository-mode-slave/6020).

 - At current version, v0.12, restarting the application is requested in order to apply changes to the [Folder Master](http://docs.syncthing.net/users/foldermaster.html) mode. Will modifications in this pull request allow to change the master/slave setting combinations "online"? As far as I've been able to understand from [the changes](https://github.com/syncthing/syncthing/pull/2417/files), this condition is not to be changed. Would it be much more laborious to achieve so?

 - *slave true* is expected to behave as if the other devices, either *master true* or *both false*, which are attached to the repository, were continuously triggering the "Override Changes" button. Would the model support not to force it? i.e., "slave true – pull-only folder, copy pushing nodes (overriding local changes if in clonflict), keep local changes that do not conflict with pushing nodes".

Both questions are related in the sense I'm using *git*'s behaviour as a reference:

- All the devices are *both false* and synced
- A device, X, is changed either to *slave true* or *both true* and multiple local changes are done.
 - If in *slave true mode*, updates from other nodes are copied, i.e. X is continuously fetching and pulling.
 - If in *both true*, X is continuously fetching, but not pulling.
- The user can change the device X either to:
  - *master true* and trigger 'Override Changes', i.e. X commits and rebases non-master-true devices
  - *both false*, i.e. X commits and pushes
  - 'Clear' button (missing, being proposed now), i.e. X removes all local changes

**Comment by Tracerneo** [comment]
>Will modifications in this pull request allow to change the master/slave setting combinations "online"? […] Would it be much more laborious to achieve so?

Folder needs to be restarted for changes to take effect, currently it's stopped at quitting Syncthing and started at turning it on again. I don't think it would be really difficult to restart only one folder, but it's not part of the PR.

>"slave true – pull-only folder, copy pushing nodes (overriding local changes if in conflict), keep local changes that do not conflict with pushing nodes".

If file exists on one device, but not on another, it **is** a conflict.
If you want to keep deleted and changed files somewhere, you want to use versioning. http://docs.syncthing.net/users/versioning.html

The git comparison seems to be right.
All folders always receive changes (remote index), regardless of mode.

Mode     | Pull | Push | Can go Out of Sync?
---------|------|------|-----
Default  | yes  | yes  | no?
Master   |      | yes  | yes
Slave    | yes  |      | never
Snapshot |      |      | always

If folder gets _out of sync_, user can trigger either Force Push (Override Changes) or Force Pull.

> 'Clear' button (missing, being proposed now), i.e. X removes all local changes

That's what Force Pull would do.

**Comment by calmh** [comment]
> user can trigger either Force Push (Override Changes) or Force Pull

That sounds acceptable :)

**Comment by Mallaco** [comment]
Force push seems like a bad idea. Unless you mean just keeping local
changes without forcefully propagating them. In this case force push might
be a bad name.

**Comment by calmh** [comment]
We already have force push, we just call the button "Override" currently.

**Comment by kbtombul** [comment]
I wonder if we need another mode for some other use cases and that could be implemented as a part of this change.

On a scenario with two devices A and B, if I want to sync the changes on device A to device B but not the other way around, I make device A Master and device B Default. This is fine until I add another device, C. If I want A and C two-way sync to each other but I don't want the changes on B to be synced to A and C, it's not that straightforward.

That's where this new mode comes in:
I'll call the mode Slave2 for the lack of creativity at this moment. The changes made to the folders in this mode are never propagated to the other devices. All the changes from other devices are pulled as long as they do NOT conflict with the local changes.

In the above scenario I can make device A and C Default, and B Slave2. This way A and C can sync freely, B can pull the changes from A and C, and if I ever make a change to B, A and C won't override.

Mode | Pull | Push | Can go Out of Sync?
-------|--------|-------|--------
Slave2 | yes | no | yes (if the local changes conflict with remote changes, also supports force pull to override local changes with remote ones)

**Comment by calmh** [comment]
I think that is the slave mode from above, at least that's how I've envisioned it.

**Comment by kbtombul** [comment]
@calmh This doesn't sound like it though (emphasis is mine):
> slave true – pull-only folder, copy pushing nodes, **always override local changes**

**Comment by calmh** [comment]
I think that's been changed to the two force options. I keep saying I'm probably not ever going to merge code that does immediate overwrite of local changes, so I hope so at least. ;)

**Comment by 1138-4EB** [comment]
> @Tracerneo: Folder needs to be restarted for changes to take effect, currently it's stopped at quitting Syncthing and started at turning it on again. I don't think it would be really difficult to restart only one folder, but it's not part of the PR.

Then, if I opened a new issue to request so, would it conflict with this PR? i.e., should I explicitly point out that it'd be better to face that after this PR is merged?

> @Tracerneo: If file exists on one device, but not on another, it is a conflict.
> If you want to keep deleted and changed files somewhere, you want to use versioning. 
> http://docs.syncthing.net/users/versioning.html

I mean to imitate the *stagging* behaviour of Git, not the history itself. However, although we are all using slightly different expression, I think that we are getting close to the same concept. The key is in @calmh's last message: *I'm probably not ever going to merge code that does immediate overwrite of local changes*.

- For the sake of simplicity, I'm assuming that all the devices belong to the same user, for user permissions not to be involved.
- The meaning of Out of Sync (OoS) in each mode:
  - Master: *defaults* have some changes not present locally.
  - Slave: there are local files not present in *defaults*.
  - Semi-detached: there are local changes not present in *defaults*. *Semi-detached* is used instead of *Slave2* used by @kbtombul.
  - Detached: there are either local changes not present in *defaults*, or changes in *defaults* which are not present locally. *Detached* is used instead of *Snapshot*, since it can always go OoS.
- *Pull*/*Push* are performed automatically. However, ther can be forced (hand-triggered) when OoS. *Override* is *Force Push*.
- *Pull*/*Push*, either automatic or forced, take *defaults* as the target/remote.

| Slave | Master | Mode     | Pull   | Push | Can go OoS? | Force Pull | Force Push | Clear | |
| ---    | ---   | ---      | ---    | ---  | ---         | ---       | ---    | --- | --- |
| 0      | 0     | Default  | D      | U    | no         | -         | -      | - | |
| 0      | 1     | Master   | -      | U    | yes           | D          |   U    | - | |
| 1      | 0     | Mirror    | D      | -    | no       |  -       |  -     | - | @Tracerneo |
| 2      | 0     | Slave    | L      | -    | yes       |  C       |  U     | C | @1138-4EB |
| 3      | 0     | Semi-detached    | P     | -    | yes       |  D       |  U     | C | @kbtombul 
| >0      | 1     | Detached | -      | -    | yes      |  D        | U      | C |

Then, the key is to differentiate regular conflicts (conflict type A - CTA) from:
 - Those produced by files which are only present locally (CTB). 
 - Those produced by files which are only present in *defaults* (CTC).

where the three groups are mutually exclusive.

The legend of the table is:

 - D (actual behaviour): update CTA + CTC locally, and remove CTB locally.
 - L: update CTA + CTC locally, but don't remove CTB.
 - P: update CTC locally, ignore CTA, and don't remove CTB.
 - U (actual behaviour): update CTA + CTB in *defaults*, and remove CTC in *defaults*.
 - C: remove CTB locally.

Alternatively, the information above may be easilly presented to the user by providing a *profiles* list, each of them containing a checkbox table. The profile corresponding to *Slave* mode would be:

|  auto   | Pull | Push | force | Pull | Push |
| --- | ---   | ---      | --- | --- | --- |
| remote   |  x    |    -     | remote | x  |  -  |
| shared  |   x   |    -    | shared | x | - |
| local      |   -   |    -     | local | x | x |

where remote is CTC, shared is CTA and local is CTB. "Clear" button could be automatically shown if CTB is not null and `Slave>0`.The default modes (*Default*, *Master*, *Mirror* and *Dettached*), could be provided by default, allowing advanced users to create new.

Now, since this PR request is about adding a boolean `Slave`, the summary above may be out of the scope. @calmh , @Tracerneo, do you prefer to keep it here? or should I open a new issue?

**Comment by calmh** [comment]
I've now spent about five minutes with my forehead in creases trying to work out what exactly the different modes are and what they accomplish. I haven't succeeded. I think it's safe to say our users wouldn't either. Can you put this in simpler terms, use fewer abbreviations and perhaps add examples for that the purpose of the different modes are to make me understand this?


---

Review status: 0 of 15 files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/syncthing/syncthing/2417)

**Comment by kbtombul** [comment]
I wonder if it's better to have options for behaviors. I'm sure it will be easier to explain what each options does than what each mode does. For example we can have 
- Pull changes automatically
- Push changes automatically
- Overwrite local changes (optionally remove locally created files?)

PS: We can move the discussion to Discourse too if you think we are going out of the scope of this pull request.

**Comment by AudriusButkevicius** [comment]
Let's get the original implementation done and merged first, and then if someone is not satisfied, they can make their own pull request implementing the mode they require. I don't want this to be a radio show where anybody calls in and asks for a mode of operation.

**Comment by 1138-4EB** [comment]
> Can you put this in simpler terms, use fewer abbreviations and perhaps add examples for what the purpose of the different modes are to make me understand this?

Sure. The following are references to the messages in this thread where I took the info from. You may go to the last block to read the summary.

- The first two columns are the fields `master` and `slave` of the config as commented by @Tracerneo in message #1.
- Columns 4 to 6 have the same meaning as in the table @Tracerneo (#6). `D` and `U` in my table refer to the actual behaviour of Syncthing, so they can be read as 'yes' in his table.
- Columns 7-8 are my try to explictly represent when would the buttons mentioned by @Tracerneo (#6) be helpful:

> If folder gets out of sync, user can trigger either Force Push (Override Changes) or Force Pull.

- Rows *Default*, *Master*, *Mirror* (he used the term 'Slave') and *Detached* (he used the term 'Snapshot') are those explained by @Tracerneo (#6). 

---

- Row *Slave* is what I asked in message #5:

> ...pull-only folder, copy pushing nodes (overriding local changes if in clonflict), keep local changes that do not conflict with pushing nodes [added now: i.e. files only present locally]

- Row *Semi-detached* is what @kbtombul proposed in message #10. He called it 'Slave2':

> The changes made to the folders in this mode are never propagated to the other devices. All the changes from other devices are pulled as long as they do NOT conflict with the local changes.

---

To sum up, is it possible to differentiate conflicts in these three mutually exclusive groups?

 - Files which are present both locally and in the devices in *Default* mode (conflic type A)
 - Files which are only present locally (conflic type B).
 - Files which are only present in the devices in *Default* mode (conflict type C).

The key difference we have explained has to do with pulling in slave (either called slave, slave2, semi-detached or mirro):
  - Conflict type A: @Tracerneo proposes keeping the remote copy. I agree. @kbtombul prefers keeping the local.
  - Conflict type B: @Tracerneo proposes removing them. @kbtombul and me think they should be kept.
  - Conflict type C: we all agree in pulling them automatically.

Does it make sense to you to let the users choose what to do with conflict types A and B when in slave mode? Would doing so signifficantly differ from what this merge is meant to? 

@AudriusButkevicius , I don't mind if you just implement what @Tracerneo commented in the first message. It's your time, and your effort. I'm just trying to get to know whether it is possible to distinguish conflict types. The table was an effort to illustrate that all the proposed variations (and others that may arise) come down to that.

**Comment by AudriusButkevicius** [comment]
So what my goal is, is to keep this mean and lean, because by the end of the day the core team will have to support it. The less knobs and whistless it has the easier for us it will be to support it, hence I don't want @Tracerneo to feel obliged to implement anything that does not relate to his usecase directly, as its unlikely it will be well though through, and well tested.

If someone is unhappy, they can make a follow up pull request satisfying their case.

**Comment by calmh** [comment]
What Audrius says, exactly.

One thing I would like to inject into this is a small terminology tweak, since Tracerneo is anyway renaming attributes, config values and GUI widgets here. I think we should leave the master/slave terminology behind and go for something that more directly describes the actions, and not in a super technical way. I'm aware that the risk for bikeshedding here is enormous (and I won't allow that to happen in this thread), but here is what I propose:

 - "Send and receive changes" (current default mode)
 - "Only send changes", maybe expressed as "Don't receive changes" (current master mode)
 - "Only receive changes" (new slave mode)

The force pull / force push things could become:

 - "Force send" (the word force here is good as it implies something forceful, in this case overwriting other peoples changes)
 - "Receive changes" (no need for the word force here, it's just the normal receive process except we only do it manually).

**Comment by dcarosone** [comment]
Send changes: Automatically Manually Now
Accept changes: Automatically Manually Now

Send might be publish, now might be "Once, now"; the point is the linking
of a manual action to the config setting, rather than the separate and
possibly confusing modes vs actions.

**Comment by generalmanager** [comment]
@calmh I wholeheartedly agree that your way of phrasing it is a huge usability improvement. Just like @lfam noticed in his proposal (https://github.com/syncthing/syncthing/issues/1439#issuecomment-120069049) every phrasing that does not break the behaviour down to a few simple statements of the actual actions taken is useless.

The way you put it is very condensed, while the meaning is unambiguous and easy to understand for non-technical users. Much better than the messy master/slave terminology!

**Comment by Tracerneo** [comment]
I won't longer work on this issue. Implementing Slave (Pull-only) folders would require either dirty hacks on hacks, pushed by even dirtier hacks, making code less maintainable…
…or rewriting half of the codebase and heavily changing the design, which would require a lot of time that I'm afraid I won't be able to find.

Current db and versioning are not well suited for updates which downgrade versions of files.
